proof/crefine/AARCH64/Fastpath_C.thy

```diff
@@ -1736,776 +1684,989 @@ lemma recv_ep_queued_st_tcb_at':
+     (wp setCTE_asidpool')
+
+lemma updateMDB_ko_at'_asidpool[wp]:
+  "\<lbrace>ko_at' (ko :: asidpool) p \<rbrace> updateMDB ptr f \<lbrace>\<lambda>_. ko_at' ko p \<rbrace>"
+  unfolding updateMDB_def Let_def
+  by (wpsimp wp: setCTE_asidpool')
+
+lemma isArchVSpacePTCap_def2:
+  "isArchVSpacePTCap cap
+   = (\<exists>base map_data. cap = ArchObjectCap (PageTableCap base VSRootPT_T map_data))"
+  by (clarsimp simp: isArchVSpacePTCap_def
+              split: capability.splits arch_capability.splits pt_type.splits)
+
+lemma capVSBasePtr_CL_capUntypedPtr_helper:
+  "\<lbrakk> ccap_relation cap cap'; isValidVTableRoot cap \<rbrakk>
+   \<Longrightarrow> capVSBasePtr_CL (cap_vspace_cap_lift cap') = global.capUntypedPtr cap"
+  by (clarsimp simp: cap_get_tag_isCap vspace_cap_capUntypedPtr_capPTBasePtr
+                     isArchVSpacePTCap_def2 ccap_relation_vspace_base
+               dest!: isValidVTableRootD')
+
+lemma casid_map_relation_Some_get_tag:
+  "casid_map_relation (Some asid_entry) asid_entry'
+   \<Longrightarrow> asid_map_get_tag asid_entry' = signed asid_map_asid_map_vspace"
+  by (clarsimp simp: casid_map_relation_def asid_map_lift_def Let_def
+               split: option.splits if_splits asid_map_CL.splits)
+
+lemma setObject_endpoint_asidpool_obj_at'[wp]:
+  "\<lbrace>obj_at' (P :: asidpool \<Rightarrow> bool) ptr\<rbrace> setObject ptr' (ep :: endpoint) \<lbrace>\<lambda>rv. obj_at' P ptr\<rbrace>"
+  apply (rule obj_at_setObject2, simp_all)
+  apply (clarsimp simp: updateObject_default_def in_monad)
+  done
+
+lemma setEndpoint_obj_at'_asidpool[wp]:
+  "\<lbrace>obj_at' (P :: asidpool \<Rightarrow> bool) t \<rbrace> setEndpoint ptr (e::endpoint) \<lbrace>\<lambda>_. obj_at' (P :: asidpool \<Rightarrow> bool) t\<rbrace>"
+  by (clarsimp simp: setEndpoint_def, wp)
+
+lemma casid_map_relation_Some_c_lift_eqs:
+  "\<lbrakk> casid_map_relation (Some asid_map_entry) asid_map_entry' \<rbrakk>
+   \<Longrightarrow> (apVMID asid_map_entry = Some vmid
+       \<longrightarrow> stored_hw_vmid_CL (asid_map_asid_map_vspace_lift asid_map_entry') = ucast vmid)
+      \<and> (stored_vmid_valid_CL (asid_map_asid_map_vspace_lift asid_map_entry') = 0)
+        = (apVMID asid_map_entry = None)
+      \<and> vspace_root_CL (asid_map_asid_map_vspace_lift asid_map_entry') = apVSpace asid_map_entry"
+  apply (clarsimp simp: casid_map_relation_def casid_map_relation_Some_get_tag
+                        to_bool_def asid_map_asid_map_vspace_lift_def
+                  split: option.splits asid_map_CL.splits)
+  apply (clarsimp simp: asid_map_lift_def Let_def ucast_ucast_mask
+                  split: if_splits)
+  done
+
+(* FIXME AARCH64 any better way? *)
+(* when we know P is true, but we can't use subst or clarsimp due to schematics *)
+lemma ccorres_If_True_drop:
+  "\<lbrakk> P; ccorres_underlying sr Gamm r xf arrel axf R R' hs a c \<rbrakk>
+   \<Longrightarrow> ccorres_underlying sr Gamm r xf arrel axf R R' hs (If P a b) c"
```

Comment:
Looks perfectly legit to me, i.e. I don't think there's a significantly better way to do this.

Reply:
removed the FIXME